dace/autodiff/backward_pass_generator.py

```diff
@@ -0,0 +1,2090 @@
+                        ONNX_AVAILABLE and isinstance(edge.dst, ONNXOp) and edge.dst_conn in attribute_to_remove
+                        for edge in out_edges):
+                    nodes_list.remove(node)
+
+    def _remove_maps_without_input_connectors(self, nodes_list: List[nodes.Node], state: SDFGState) -> None:
+        """Remove maps that don't have any input connectors from the nodes_list.
+
+        These are maps that won't have an output in the backward pass and thus can be skipped from the reversal process.
+        Note that we do not remove the AccessNode that the no-input map writes to.
+        This is because we might need to zero out the gradient of this node.
+        If no zeroing out is necessary, the node will be removed in the reverse_subgraph function cleanup at the end.
+        """
+        for node in nodes_list[:]:  # Iterate over a copy of the list to avoid modification issues
+            if isinstance(node, nodes.MapEntry) and len(node.in_connectors) == 0:
+                nodes_list.remove(node)
+                # Remove the MapExit and everything in between
+                # Get the equivalent map exit for the map entry
+                map_exit = state.exit_node(node)
+                nodes_list.remove(map_exit)
+
+                # Get all the nodes between the map entry and exit
+                for state_node in state.nodes():
+                    # Check the scope of the node if it is within the map
+                    if state_node in state.scope_dict() and state.scope_dict(
+                    )[state_node] == node and state_node in nodes_list:
+                        nodes_list.remove(state_node)
+
+    def _find_subgraph_to_differentiate(self) -> None:
+        """Determine which nodes we need to reverse; this forms the subgraph we will differentiate.
+
+        We do a reverse BFS from the target output node.
+        In the case where a state is within a loop, this may result in different subgraphs
+        depending on the loop iteration.
+
+        To calculate the gradients for a node x in ``required_gradients``, we need to sum up the gradient
+        contributions from every node y where x is used as an input.
+        """
+        backward_nodes: set[nodes.Node] = set()
+        given_gradients_all_states = set(self.given_gradients_data)
+
+        required_gradients_all_states = {n for n in self.required_gradients_data}
+        given_gradients_all_states = given_gradients_all_states | required_gradients_all_states
+
+        # Do the backward BFS iterativly
+        for state in reversed(self.state_order):
+            state_given_gradients: List[nodes.AccessNode] = []
+
+            for node in state:
+                if isinstance(node, nodes.AccessNode) and node.data in given_gradients_all_states:
+                    state_given_gradients.append(node)
+
+            backward_nodes = {n for e in state.edge_bfs(state_given_gradients, reverse=True) for n in [e.src, e.dst]}
```

Comment:
Why do you add both `e.src` and `e.dst` here? Is it so that `state_given_gradients` is also added to the set?